pkg/splunk/enterprise/indexercluster.go

```diff
@@ -1067,6 +1094,142 @@ func RetrieveCMSpec(ctx context.Context, client splcommon.ControllerClient, cr *
+	eventPublisher, _ := newK8EventPublisher(c, cr)
+
+	// get the clusterManagerRef attached to the instance
+	clusterManagerRef := cr.Spec.ClusterManagerRef
+
+	if mgr.c == nil {
+		mgr.c = c
+	}
+
+	cm := mgr.getClusterManagerClient(ctx)
+	clusterInfo, err := cm.GetClusterInfo(false)
+	if err != nil {
+		return false, fmt.Errorf("could not get cluster info from cluster manager. Error=%s", err.Error())
+	}
+	if clusterInfo.MultiSite == "true" {
+		opts := []rclient.ListOption{
+			rclient.InNamespace(cr.GetNamespace()),
+		}
+		indexerList, err := getIndexerClusterList(ctx, c, cr, opts)
+		if err != nil {
+			return false, err
+		}
+		sortedList, err := getIndexerClusterSortedSiteList(ctx, c, cr.Spec.ClusterManagerRef, indexerList)
+
+		preIdx := enterpriseApi.IndexerCluster{}
+
+		for i, v := range sortedList.Items {
+			if &v == cr {
+				if i > 0 {
+					preIdx = sortedList.Items[i-1]
+				}
+				break
+
+			}
+		}
+		if len(preIdx.Name) != 0 {
+			image, _ := getCurrentImage(ctx, c, &preIdx, SplunkIndexer)
+			if preIdx.Status.Phase != enterpriseApi.PhaseReady || image != cr.Spec.Image {
+				return false, nil
+			}
+		}
+
+	}
+
+	// check if a search head cluster exists with the same ClusterManager instance attached
+	searchHeadClusterInstance := enterpriseApi.SearchHeadCluster{}
+	opts := []rclient.ListOption{
+		rclient.InNamespace(cr.GetNamespace()),
+	}
+	searchHeadList, err := getSearchHeadClusterList(ctx, c, cr, opts)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return true, nil
+		}
+		return false, err
+	}
+	if len(searchHeadList.Items) == 0 {
+		return true, nil
+	}
+
+	// check if instance has the required ClusterManagerRef
+	for _, shc := range searchHeadList.Items {
+		if shc.Spec.ClusterManagerRef.Name == clusterManagerRef.Name {
+			searchHeadClusterInstance = shc
+			break
+		}
+	}
+	if len(searchHeadClusterInstance.GetName()) == 0 {
+		return true, nil
+	}
+
+	shcImage, err := getCurrentImage(ctx, c, &searchHeadClusterInstance, SplunkSearchHead)
+	if err != nil {
+		eventPublisher.Warning(ctx, "isIndexerClusterReadyForUpgrade", fmt.Sprintf("Could not get the Search Head Cluster Image. Reason %v", err))
+		scopedLog.Error(err, "Unable to get SearchHeadCluster current image")
+		return false, err
+	}
+
+	idxImage, err := getCurrentImage(ctx, c, cr, SplunkIndexer)
+	if err != nil {
+		eventPublisher.Warning(ctx, "isIndexerClusterReadyForUpgrade", fmt.Sprintf("Could not get the Indexer Cluster Image. Reason %v", err))
+		scopedLog.Error(err, "Unable to get IndexerCluster current image")
+		return false, err
+	}
+
+	// check if an image upgrade is happening and whether SHC has finished updating yet, return false to stop
+	// further reconcile operations on IDX until SHC is ready
+	if (cr.Spec.Image != idxImage) && (searchHeadClusterInstance.Status.Phase != enterpriseApi.PhaseReady || shcImage != cr.Spec.Image) {
```

Comment:
Given the whole upgrade process, we may need to add in the documentation that we **expect** all Splunk CRs in the namespace/otherwise need to be on the same image/version.